### PR TITLE
[Dev] Fix DSA sparse attention backward empty rows

### PR DESCRIPTION
**Pull request by buptzyb**
- [x] I, the PR author, have personally reviewed every line of this PR.

# What does this PR do?

Supersedes #6910, which GitHub automatically closed after its temporary base
branch was deleted. Main counterpart: #6939.

This change builds on the DSA pre-attention infrastructure merged in #6206 and
is directly based on `dev`.

Keep ordinary DSA sparse-attention backward shapes static when top-k contains
empty query rows, removing its dynamic `nonzero`/compact/gather/scatter path.

The generic path promotes each empty row to a safe one-tile cuDNN invocation:
its first index is zero and its effective top-k length is one. FlashMLA's
supported empty-row contract produces `O=0` and `LSE=+inf`; before backward,
the wrapper masks `dO` to zero and sets `LSE` to `+inf`, making
`P = exp(S - LSE) = 0`. The returned `dQ` is explicitly zeroed for those rows,
so the safe tile contributes neither query nor KV gradients. A conservative
host-side certificate retains the allocation-free fast path when all top-k
rows are provably nonempty.

This covers both the combined fused path and the split path used by IndexShare
and MTP sharing. It does not change CSA, THD/CP layout, MTP sharing semantics,
configuration, or the full-iteration CUDA Graph validator.

## Issue tracking

Linked issue: N/A

### Pre-checks

- [x] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [x] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [x] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Verification notes

- Post-rebase targeted unit tests: 20 core cases passed, covering full and split
  paths; nonempty, mixed-empty, and all-empty rows; BF16/FP32; metadata
  immutability; packed-CP padding; batch isolation; and the plain-causal and
  local-varlen certificate paths.
- Full hook-compatibility plus zero-size rejection selection: 13 passed. Across
  both selections, 31 unique cases passed.
- Megatron-LM autoformat check, pylint (10.00/10), and ruff passed against
  `dev`.

### CODE REVIEW

File: megatron/core/transformer/experimental_attention_variant/dsa_cudnn_kernels.py [hunk 1]
```diff
@@ -2438,56 +2442,31 @@ def _run_sparse_attention_backward(
...
-    global_idxs = torch.cat((global_idxs, torch.zeros_like(global_idxs[:1])), dim=0)
-    topk_length = torch.cat((topk_length, torch.ones_like(topk_length[:1])), dim=0)
+    empty_rows = None
+    if not all_topk_rows_nonempty:
+        empty_rows = topk_length <= 0
+        topk_length = topk_length.masked_fill(empty_rows, 1)
+        bwd_dO_flat = bwd_dO_flat.masked_fill(empty_rows[:, None, None], 0)
+        # FlashMLA already returns zero output and LSE=+inf for empty rows. Preserve
+        # that LSE invariant after promoting them to one safe tile so probabilities stay zero.
+        bwd_lse = bwd_lse.masked_fill(empty_rows[:, None], float("inf"))
```
Comment (yaox12) [inline megatron/core/transformer/experimental_attention_variant/dsa_cudnn_kernels.py]: **[IMPORTANT Verification]** Empty rows with `LSE=+inf` are now fed to the real cuDNN `sparse_attention_backward_wrapper`. Before this PR those rows were compacted out, so the production kernel never saw an infinite LSE.

Whether the CuTe kernel computes `exp(S - LSE)` cleanly (→ 0) or hits an `inf - inf` / `inf * 0` NaN depends on its internal formulation (log2-scaled LSE, max-logit tricks, masked slots at `-inf`, etc.). Every new test in this PR stubs the kernel with `FakeDSA`, and the real-kernel tests in this file are disabled in CI, so this runtime assumption is currently untested. A NaN here would poison dKV for the whole batch through the shared safe tile at global row 0.

**Suggestion:** add a real-kernel check (behind the existing cuDNN/FlashMLA skip conditions) that runs a mixed-empty-row BF16 case through the true wrapper and asserts `torch.isfinite` on dQ/dKV plus zero dKV contribution from the empty rows, or state in the Verification notes that this was validated on hardware. It would also be good to confirm with the cuDNN owners that `+inf` LSE is within the wrapper's documented input contract.

File: megatron/core/transformer/experimental_attention_variant/dsa_masking.py [hunk 2]
```diff
@@ -33,6 +34,46 @@
...
+    packed self-attention layout. Selected cumulative lengths (padded when available),
+    synthetic tail positions, and reordered KV cover every physical query row; checking
+    tensor values here would introduce a device sync. This proves geometric non-emptiness
+    only; real-token masking remains a separate concern.
+
+    Backend callers may pass their normalized plain-causal layout as
+    ``varlen_is_plain_causal=True`` after dropping the equivalent explicit bounds.
+    """
+    if (
+        not computes_topk
```
Comment (yaox12) [inline megatron/core/transformer/experimental_attention_variant/dsa_masking.py:62]: **[SUGGESTION Performance]** The certificate returns `False` whenever `computes_topk` is `False`, so every IndexShare / MTP-sharing consumer layer always takes the generic backward path, as do CP>1 non-packed causal runs (`varlen_is_plain_causal` is `False` under CP).

The generic path materializes three full copies per layer per backward (`dO`, `LSE`, `topk_length`). For consumer layers this is pure overhead: the producer layer already proved the shared top-k non-empty and the consumer uses the same indices. At 60+ layers on DSv3.2-scale models these are non-trivial extra allocations on the backward critical path.

**Suggestion (fine as a follow-up / TODO):** propagate the producer's `all_topk_rows_nonempty` alongside the shared top-k indices/lengths so consumers inherit it, and consider extending the certificate to the CP allgather causal layout (each local query row has global position ≥ its local index, so it always has at least its own key).

File: megatron/core/transformer/experimental_attention_variant/dsa_cudnn_kernels.py [hunk 3]
```diff
@@ -2438,56 +2442,31 @@ def _run_sparse_attention_backward(
...
-    bwd_out_flat = torch.cat((bwd_out_flat, torch.zeros_like(bwd_out_flat[:1])), dim=0)
-    bwd_dO_flat = torch.cat((bwd_dO_flat, torch.zeros_like(bwd_dO_flat[:1])), dim=0)
-    bwd_lse = torch.cat((bwd_lse, torch.zeros_like(bwd_lse[:1])), dim=0)
-    global_idxs = torch.cat((global_idxs, torch.zeros_like(global_idxs[:1])), dim=0)
-    topk_length = torch.cat((topk_length, torch.ones_like(topk_length[:1])), dim=0)
+    empty_rows = None
+    if not all_topk_rows_nonempty:
+        empty_rows = topk_length <= 0
+        topk_length = topk_length.masked_fill(empty_rows, 1)
+        bwd_dO_flat = bwd_dO_flat.masked_fill(empty_rows[:, None, None], 0)
```
Comment (yaox12) [inline megatron/core/transformer/experimental_attention_variant/dsa_cudnn_kernels.py]: **[SUGGESTION Memory]** `bwd_dO_flat.masked_fill(...)` and `bwd_lse.masked_fill(...)` are out-of-place even when the head-padding branch above just allocated `bwd_dO_flat` / `bwd_lse` as fresh buffers, so in the padded-head case dO is copied twice (padded copy + masked copy).

**Suggestion:** use `masked_fill_` when `padded_num_heads != num_heads` (buffers are owned), and keep the out-of-place form otherwise since `grad_output` and the saved `lse` must not be mutated.

File: megatron/core/transformer/experimental_attention_variant/dsa_cudnn_kernels.py (hunk 1 above)
Comment (buptzyb) [reply megatron/core/transformer/experimental_attention_variant/dsa_cudnn_kernels.py]: Added a real cuDNN BF16 regression. It covers mixed valid/empty rows with LSE=+inf, compares against a valid-only baseline, and checks finite dQ/dKV plus no dKV contamination.

File: megatron/core/transformer/experimental_attention_variant/dsa_masking.py (hunk 2 above)
Comment (buptzyb) [reply megatron/core/transformer/experimental_attention_variant/dsa_masking.py:62]: Good point! I’ll leave this as a follow-up after #6472 lands, since ForwardSharingState is the right place to carry the certificate together with the shared top-k metadata. The current generic path remains correct.

File: megatron/core/transformer/experimental_attention_variant/dsa_cudnn_kernels.py (hunk 3 above)
Comment (buptzyb) [reply megatron/core/transformer/experimental_attention_variant/dsa_cudnn_kernels.py]: Done! The padded dO/LSE buffers now use masked_fill_, while the non-padded path stays out-of-place to avoid mutating grad_output or the saved LSE.
